pkg/collector/corechecks/gpu/nvmlmetrics/collector.go

```diff
@@ -0,0 +1,161 @@
+func NewCollector(lib nvml.Interface) (*Collector, error) {
+	return newCollectorWithSubsystems(lib, allSubsystems)
+}
+
+// newCollectorWithSubsystems allows specifying which subsystems to use when creating the collector, useful for tests.
+func newCollectorWithSubsystems(lib nvml.Interface, subsystems map[string]subsystemFactory) (*Collector, error) {
+	coll := &Collector{
+		lib:        lib,
+		collectors: make(map[nvml.Device][]subsystemCollector),
+	}
+
+	devCount, ret := lib.DeviceGetCount()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("failed to get device count: %s", nvml.ErrorString(ret))
+	}
+
+	for i := 0; i < devCount; i++ {
+		dev, ret := coll.lib.DeviceGetHandleByIndex(i)
+		if ret != nvml.SUCCESS {
+			return nil, fmt.Errorf("failed to get device handle for index %d: %s", i, nvml.ErrorString(ret))
+		}
+
+		for name, factory := range subsystems {
+			subsystem, err := factory(lib, dev)
+			if errors.Is(err, errUnsupportedDevice) {
+				log.Warnf("device %s does not support collector %s", dev, name)
+				continue
+			} else if err != nil {
+				_ = coll.Close() // Close all previously created subsystems
+				return nil, fmt.Errorf("failed to create subsystem %s: %w", name, err)
+			}
+
+			coll.collectors[dev] = append(coll.collectors[dev], subsystem)
+		}
+	}
+
+	return coll, nil
+}
+
+// Collect collects metrics from all the subsystems and returns them. It will try to return as many
+// metrics as possible, even if some subsystems fail to collect metrics. This means that even if the
+// error return is not nil, the returned metrics might still be useful.
+func (coll *Collector) Collect() ([]Metric, error) {
+	var allMetrics []Metric
+	var err error
+
+	for dev, collectors := range coll.collectors {
+		tags, tagsErr := getTagsFromDevice(dev)
+		if tagsErr != nil {
+			return allMetrics, fmt.Errorf("failed to get tags for device: %w", tagsErr)
+		}
+
+		for _, subsystem := range collectors {
+			metrics, collectErr := subsystem.collect()
+			if collectErr != nil {
+				err = multierror.Append(err, fmt.Errorf("failed to collect metrics for subsystem %s: %w", subsystem.name(), collectErr))
+			}
+
+			for _, metric := range metrics {
+				metric.Tags = append(metric.Tags, tags...)
+				allMetrics = append(allMetrics, metric)
+			}
+		}
+	}
+
+	return allMetrics, err
+}
+
+// Close closes the collector and releases any resources it might have allocated.
+func (coll *Collector) Close() error {
+	var errs error
+
+	for _, collectors := range coll.collectors {
+		for _, subsystem := range collectors {
+			if err := subsystem.close(); err != nil {
+				errs = multierror.Append(errs, err)
+			}
+		}
+	}
+
+	return errs
+}
+
+// getTagsFromDevice returns the tags associated with the given NVML device.
+func getTagsFromDevice(dev nvml.Device) ([]string, error) {
+	uuid, ret := dev.GetUUID()
+	if ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("failed to get device UUID: %s", nvml.ErrorString(ret))
```

Comment:
i am not sure you want to omit ALL the tags if you can't get one. I think tags are usually attached as a  best effort in datadog-agent

Reply:
Changed, returning on best-effort.